Generate OpenAPI specification and Swagger file

jinchihe opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

The PR is going to generate OpenAPI specification, and  generate Swagger file according to the OpenAPI specification.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added): 

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes
```
None
```
<!--
Describe any user facing changes here, or delete this block.

Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior
- Changes requiring upgrade notices or deprecation warnings

For pull requests with a release note:

    ```release-note
    Your release note here
    ```

For pull requests that require additional action from users switching to the new release, include the string "action required" (case insensitive) in the release note:

    ```release-note
    action required: your release note here
    ```

For pull requests that don't need to be mentioned at release time, use the `/release-note-none` Prow command to add the `release-note-none` label to the PR. You can also write the string "NONE" as a release note in your PR description:

    ```release-note
    None
    ```
-->

jinchihe commented:
/cc @vdemeester

vdemeester commented:
@jinchihe thanks for the PR :angel: 

> Firstly, I'm not sure that's good place to put the SDK here, personally it's better to have a separate repo to maintain SDK, such as python and java (if needed) etc... I'm waiting for comments from maintainers for this.

I tend to agree, we could have the SDK in a separate repository (a bit like `k8s.io/client-go`).

Overall, I think it look good, but I feel we should split the PR into several : 
1. adding openapi gen (go code and schemas) in one PR
2. swagger in another
3. the sdk generation in a third and/or in a separate repository :wink: 

/cc @afrittoli @bobcatfish @sbwsg

bobcatfish commented:
Would it make sense to have keep the python client lib generation as a separate project, maybe in experimental or a separate repo? @tektoncd/core-maintainers

ImJasonH commented:
> Would it make sense to have keep the python client lib generation as a separate project, maybe in experimental or a separate repo? @tektoncd/core-maintainers

I would prefer that. Either as a directory in tektoncd/experimental, or just as a separately-owned repo outside the Tekton org, based on Tekton-owned generated Swagger.

Basically, I definitely think Tekton should own Swagger doc generation, but I don't want to own generated Python (or any non-Go) client libraries since I don't think we can rely on sufficient Python expertise in case there are issues.

jinchihe commented:
> I tend to agree, we could have the SDK in a separate repository (a bit like `k8s.io/client-go`).
> 
> Overall, I think it look good, but I feel we should split the PR into several :
> 
> 1. adding openapi gen (go code and schemas) in one PR
> 2. swagger in another
> 3. the sdk generation in a third and/or in a separate repository 😉

@vdemeester Thank you for the comments. Make sense,  we may can commit the openapi and swagger in `tektoncd/pipeline` repo, and commit sdk after the separate repository (may be named `tektoncd/sdk`) created.

jinchihe commented:
> I would prefer that. Either as a directory in tektoncd/experimental, or just as a separately-owned repo outside the Tekton org, based on Tekton-owned generated Swagger.
> 
> Basically, I definitely think Tekton should own Swagger doc generation, but I don't want to own generated Python (or any non-Go) client libraries since I don't think we can rely on sufficient Python expertise in case there are issues.

@ImJasonH why put outside the Tekton org? personally I suggest to maintain that inside of Tekton org, you know, that has close relationship with tekton/pipeline and very import for  tekton end user. If you‘re worry about sufficient Python expertise, this is open source, I think if have problems, more and more Python expertise will take part in :-)

ImJasonH commented:
> @ImJasonH why put outside the Tekton org? personally I suggest to maintain that inside of Tekton org, you know, that has close relationship with tekton/pipeline and very import for tekton end user. If you‘re worry about sufficient Python expertise, this is open source, I think if have problems, more and more Python expertise will take part in :-)

Perhaps I should rephrase. 😄  I'd like to _start_ with the generated Python client (and other generated clients) in a non-Tekton repo, or in `tektoncd/experimental` where user expectations around ownership and support are more relaxed. If there's sufficient interest and usage of the clients in the future, and if an interested owner steps up, we could promote it to a separate `tektoncd/*` repo, or possibly eventually into `tektoncd/pipeline` -- but I don't think we should start there.

Having generated clients means having a release process for them, publishing them to PyPi, etc., and I just don't think there's enough dedicated expertise or interest among Tekton owners and partners at this time to justify adding that burden. If the clients are a hit and are widely used, then presumably some dependent user would be willing to step and own generating and releasing new clients, and debugging the inevitable user issues. The best way to gauge use and interest is to start by publishing them -- _anywhere_ -- and seeing how they get adopted.

jinchihe commented:
@ImJasonH @vdemeester OK thanks. I will separate the PR to keep the openapi and swagger parts in this repo, and commit the generated python SDK in `tektoncd/experimental`.

jinchihe commented:
@ImJasonH @vdemeester I just reconsidered that, seems the swagger parts are useless in the `tektoncd/pipeline` repo, but that's path to generate SDKs for Tekton. so I just keep the openapi part in this repo, and move the swagger and generated SDK to `tektoncd/experimental`. If you have comments please let me konw.

if so, the PR is ready for reviewing and testing. I will update summary and description. Thanks.

vdemeester commented:
/ok-to-test
/kind feature

jinchihe commented:
Regenerated since the Sidecar feature. Thanks.

jinchihe commented:
@vdemeester Any more comments? Thanks

jinchihe commented:
>  just one ask is to add ./hack/update-openapigen.sh to ./hack/update-codegen.sh

@vdemeester Great thanks.  Updated.

vdemeester commented:
```diff
Diffing /home/prow/go/src/github.com/tektoncd/pipeline against freshly generated codegen
diff -Naupr /home/prow/go/src/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1/openapi_generated.go /home/prow/go/src/github.com/tektoncd/pipeline/tmpdiffroot.KQ9eiZ/pkg/apis/pipeline/v1beta1/openapi_generated.go
--- /home/prow/go/src/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1/openapi_generated.go	2020-08-18 02:58:33.396260921 +0000
+++ /home/prow/go/src/github.com/tektoncd/pipeline/tmpdiffroot.KQ9eiZ/pkg/apis/pipeline/v1beta1/openapi_generated.go	2020-08-18 02:48:17.853123990 +0000
@@ -29,78 +29,78 @@ import (
 
 func GetOpenAPIDefinitions(ref common.ReferenceCallback) map[string]common.OpenAPIDefinition {
 	return map[string]common.OpenAPIDefinition{
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/pod.Template":                              schema_pkg_apis_pipeline_pod_Template(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.ArrayOrString":                     schema_pkg_apis_pipeline_v1beta1_ArrayOrString(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.CannotConvertError":                schema_pkg_apis_pipeline_v1beta1_CannotConvertError(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.CloudEventDelivery":                schema_pkg_apis_pipeline_v1beta1_CloudEventDelivery(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.CloudEventDeliveryState":           schema_pkg_apis_pipeline_v1beta1_CloudEventDeliveryState(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.ClusterTask":                       schema_pkg_apis_pipeline_v1beta1_ClusterTask(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.ClusterTaskList":                   schema_pkg_apis_pipeline_v1beta1_ClusterTaskList(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.ConditionCheck":                    schema_pkg_apis_pipeline_v1beta1_ConditionCheck(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.ConditionCheckStatus":              schema_pkg_apis_pipeline_v1beta1_ConditionCheckStatus(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.ConditionCheckStatusFields":        schema_pkg_apis_pipeline_v1beta1_ConditionCheckStatusFields(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.InternalTaskModifier":              schema_pkg_apis_pipeline_v1beta1_InternalTaskModifier(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.Param":                             schema_pkg_apis_pipeline_v1beta1_Param(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.ParamSpec":                         schema_pkg_apis_pipeline_v1beta1_ParamSpec(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.Pipeline":                          schema_pkg_apis_pipeline_v1beta1_Pipeline(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineDeclaredResource":          schema_pkg_apis_pipeline_v1beta1_PipelineDeclaredResource(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineList":                      schema_pkg_apis_pipeline_v1beta1_PipelineList(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineRef":                       schema_pkg_apis_pipeline_v1beta1_PipelineRef(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineResourceBinding":           schema_pkg_apis_pipeline_v1beta1_PipelineResourceBinding(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineResourceRef":               schema_pkg_apis_pipeline_v1beta1_PipelineResourceRef(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineResourceResult":            schema_pkg_apis_pipeline_v1beta1_PipelineResourceResult(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineResult":                    schema_pkg_apis_pipeline_v1beta1_PipelineResult(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineRun":                       schema_pkg_apis_pipeline_v1beta1_PipelineRun(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineRunConditionCheckStatus":   schema_pkg_apis_pipeline_v1beta1_PipelineRunConditionCheckStatus(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineRunList":                   schema_pkg_apis_pipeline_v1beta1_PipelineRunList(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineRunResult":                 schema_pkg_apis_pipeline_v1beta1_PipelineRunResult(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineRunSpec":                   schema_pkg_apis_pipeline_v1beta1_PipelineRunSpec(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineRunSpecServiceAccountName": schema_pkg_apis_pipeline_v1beta1_PipelineRunSpecServiceAccountName(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineRunStatus":                 schema_pkg_apis_pipeline_v1beta1_PipelineRunStatus(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineRunStatusFields":           schema_pkg_apis_pipeline_v1beta1_PipelineRunStatusFields(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineRunTaskRunStatus":          schema_pkg_apis_pipeline_v1beta1_PipelineRunTaskRunStatus(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineSpec":                      schema_pkg_apis_pipeline_v1beta1_PipelineSpec(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineTask":                      schema_pkg_apis_pipeline_v1beta1_PipelineTask(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineTaskCondition":             schema_pkg_apis_pipeline_v1beta1_PipelineTaskCondition(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineTaskInputResource":         schema_pkg_apis_pipeline_v1beta1_PipelineTaskInputResource(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineTaskOutputResource":        schema_pkg_apis_pipeline_v1beta1_PipelineTaskOutputResource(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineTaskParam":                 schema_pkg_apis_pipeline_v1beta1_PipelineTaskParam(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineTaskResources":             schema_pkg_apis_pipeline_v1beta1_PipelineTaskResources(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineTaskRun":                   schema_pkg_apis_pipeline_v1beta1_PipelineTaskRun(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineTaskRunSpec":               schema_pkg_apis_pipeline_v1beta1_PipelineTaskRunSpec(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.PipelineWorkspaceDeclaration":      schema_pkg_apis_pipeline_v1beta1_PipelineWorkspaceDeclaration(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.ResultRef":                         schema_pkg_apis_pipeline_v1beta1_ResultRef(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.Sidecar":                           schema_pkg_apis_pipeline_v1beta1_Sidecar(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.SidecarState":                      schema_pkg_apis_pipeline_v1beta1_SidecarState(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.Step":                              schema_pkg_apis_pipeline_v1beta1_Step(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.StepState":                         schema_pkg_apis_pipeline_v1beta1_StepState(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.Task":                              schema_pkg_apis_pipeline_v1beta1_Task(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskList":                          schema_pkg_apis_pipeline_v1beta1_TaskList(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskRef":                           schema_pkg_apis_pipeline_v1beta1_TaskRef(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskResource":                      schema_pkg_apis_pipeline_v1beta1_TaskResource(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskResourceBinding":               schema_pkg_apis_pipeline_v1beta1_TaskResourceBinding(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskResources":                     schema_pkg_apis_pipeline_v1beta1_TaskResources(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskResult":                        schema_pkg_apis_pipeline_v1beta1_TaskResult(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskRun":                           schema_pkg_apis_pipeline_v1beta1_TaskRun(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskRunInputs":                     schema_pkg_apis_pipeline_v1beta1_TaskRunInputs(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskRunList":                       schema_pkg_apis_pipeline_v1beta1_TaskRunList(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskRunOutputs":                    schema_pkg_apis_pipeline_v1beta1_TaskRunOutputs(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskRunResources":                  schema_pkg_apis_pipeline_v1beta1_TaskRunResources(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskRunResult":                     schema_pkg_apis_pipeline_v1beta1_TaskRunResult(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskRunSpec":                       schema_pkg_apis_pipeline_v1beta1_TaskRunSpec(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskRunStatus":                     schema_pkg_apis_pipeline_v1beta1_TaskRunStatus(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskRunStatusFields":               schema_pkg_apis_pipeline_v1beta1_TaskRunStatusFields(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.TaskSpec":                          schema_pkg_apis_pipeline_v1beta1_TaskSpec(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.WorkspaceBinding":                  schema_pkg_apis_pipeline_v1beta1_WorkspaceBinding(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.WorkspaceDeclaration":              schema_pkg_apis_pipeline_v1beta1_WorkspaceDeclaration(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.WorkspacePipelineTaskBinding":      schema_pkg_apis_pipeline_v1beta1_WorkspacePipelineTaskBinding(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/resource/v1alpha1.PipelineResource":                 schema_pkg_apis_resource_v1alpha1_PipelineResource(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/resource/v1alpha1.PipelineResourceList":             schema_pkg_apis_resource_v1alpha1_PipelineResourceList(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/resource/v1alpha1.PipelineResourceSpec":             schema_pkg_apis_resource_v1alpha1_PipelineResourceSpec(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/resource/v1alpha1.PipelineResourceStatus":           schema_pkg_apis_resource_v1alpha1_PipelineResourceStatus(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/resource/v1alpha1.ResourceDeclaration":              schema_pkg_apis_resource_v1alpha1_ResourceDeclaration(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/resource/v1alpha1.ResourceParam":                    schema_pkg_apis_resource_v1alpha1_ResourceParam(ref),
-		"github.com/tektoncd/pipeline/pkg/apis/resource/v1alpha1.SecretParam":                      schema_pkg_apis_resource_v1alpha1_SecretParam(ref),
+		"./pkg/apis/pipeline/pod.Template":                              schema_pkg_apis_pipeline_pod_Template(ref),
+		"./pkg/apis/pipeline/v1beta1.ArrayOrString":                     schema_pkg_apis_pipeline_v1beta1_ArrayOrString(ref),
+		"./pkg/apis/pipeline/v1beta1.CannotConvertError":                schema_pkg_apis_pipeline_v1beta1_CannotConvertError(ref),
+		"./pkg/apis/pipeline/v1beta1.CloudEventDelivery":                schema_pkg_apis_pipeline_v1beta1_CloudEventDelivery(ref),
+		"./pkg/apis/pipeline/v1beta1.CloudEventDeliveryState":           schema_pkg_apis_pipeline_v1beta1_CloudEventDeliveryState(ref),
+		"./pkg/apis/pipeline/v1beta1.ClusterTask":                       schema_pkg_apis_pipeline_v1beta1_ClusterTask(ref),
+		"./pkg/apis/pipeline/v1beta1.ClusterTaskList":                   schema_pkg_apis_pipeline_v1beta1_ClusterTaskList(ref),
+		"./pkg/apis/pipeline/v1beta1.ConditionCheck":                    schema_pkg_apis_pipeline_v1beta1_ConditionCheck(ref),
+		"./pkg/apis/pipeline/v1beta1.ConditionCheckStatus":              schema_pkg_apis_pipeline_v1beta1_ConditionCheckStatus(ref),
+		"./pkg/apis/pipeline/v1beta1.ConditionCheckStatusFields":        schema_pkg_apis_pipeline_v1beta1_ConditionCheckStatusFields(ref),
+		"./pkg/apis/pipeline/v1beta1.InternalTaskModifier":              schema_pkg_apis_pipeline_v1beta1_InternalTaskModifier(ref),
+		"./pkg/apis/pipeline/v1beta1.Param":                             schema_pkg_apis_pipeline_v1beta1_Param(ref),
+		"./pkg/apis/pipeline/v1beta1.ParamSpec":                         schema_pkg_apis_pipeline_v1beta1_ParamSpec(ref),
+		"./pkg/apis/pipeline/v1beta1.Pipeline":                          schema_pkg_apis_pipeline_v1beta1_Pipeline(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineDeclaredResource":          schema_pkg_apis_pipeline_v1beta1_PipelineDeclaredResource(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineList":                      schema_pkg_apis_pipeline_v1beta1_PipelineList(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineRef":                       schema_pkg_apis_pipeline_v1beta1_PipelineRef(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineResourceBinding":           schema_pkg_apis_pipeline_v1beta1_PipelineResourceBinding(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineResourceRef":               schema_pkg_apis_pipeline_v1beta1_PipelineResourceRef(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineResourceResult":            schema_pkg_apis_pipeline_v1beta1_PipelineResourceResult(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineResult":                    schema_pkg_apis_pipeline_v1beta1_PipelineResult(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineRun":                       schema_pkg_apis_pipeline_v1beta1_PipelineRun(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineRunConditionCheckStatus":   schema_pkg_apis_pipeline_v1beta1_PipelineRunConditionCheckStatus(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineRunList":                   schema_pkg_apis_pipeline_v1beta1_PipelineRunList(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineRunResult":                 schema_pkg_apis_pipeline_v1beta1_PipelineRunResult(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineRunSpec":                   schema_pkg_apis_pipeline_v1beta1_PipelineRunSpec(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineRunSpecServiceAccountName": schema_pkg_apis_pipeline_v1beta1_PipelineRunSpecServiceAccountName(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineRunStatus":                 schema_pkg_apis_pipeline_v1beta1_PipelineRunStatus(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineRunStatusFields":           schema_pkg_apis_pipeline_v1beta1_PipelineRunStatusFields(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineRunTaskRunStatus":          schema_pkg_apis_pipeline_v1beta1_PipelineRunTaskRunStatus(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineSpec":                      schema_pkg_apis_pipeline_v1beta1_PipelineSpec(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineTask":                      schema_pkg_apis_pipeline_v1beta1_PipelineTask(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineTaskCondition":             schema_pkg_apis_pipeline_v1beta1_PipelineTaskCondition(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineTaskInputResource":         schema_pkg_apis_pipeline_v1beta1_PipelineTaskInputResource(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineTaskOutputResource":        schema_pkg_apis_pipeline_v1beta1_PipelineTaskOutputResource(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineTaskParam":                 schema_pkg_apis_pipeline_v1beta1_PipelineTaskParam(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineTaskResources":             schema_pkg_apis_pipeline_v1beta1_PipelineTaskResources(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineTaskRun":                   schema_pkg_apis_pipeline_v1beta1_PipelineTaskRun(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineTaskRunSpec":               schema_pkg_apis_pipeline_v1beta1_PipelineTaskRunSpec(ref),
+		"./pkg/apis/pipeline/v1beta1.PipelineWorkspaceDeclaration":      schema_pkg_apis_pipeline_v1beta1_PipelineWorkspaceDeclaration(ref),
+		"./pkg/apis/pipeline/v1beta1.ResultRef":                         schema_pkg_apis_pipeline_v1beta1_ResultRef(ref),
+		"./pkg/apis/pipeline/v1beta1.Sidecar":                           schema_pkg_apis_pipeline_v1beta1_Sidecar(ref),
+		"./pkg/apis/pipeline/v1beta1.SidecarState":                      schema_pkg_apis_pipeline_v1beta1_SidecarState(ref),
+		"./pkg/apis/pipeline/v1beta1.Step":                              schema_pkg_apis_pipeline_v1beta1_Step(ref),
+		"./pkg/apis/pipeline/v1beta1.StepState":                         schema_pkg_apis_pipeline_v1beta1_StepState(ref),
+		"./pkg/apis/pipeline/v1beta1.Task":                              schema_pkg_apis_pipeline_v1beta1_Task(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskList":                          schema_pkg_apis_pipeline_v1beta1_TaskList(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskRef":                           schema_pkg_apis_pipeline_v1beta1_TaskRef(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskResource":                      schema_pkg_apis_pipeline_v1beta1_TaskResource(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskResourceBinding":               schema_pkg_apis_pipeline_v1beta1_TaskResourceBinding(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskResources":                     schema_pkg_apis_pipeline_v1beta1_TaskResources(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskResult":                        schema_pkg_apis_pipeline_v1beta1_TaskResult(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskRun":                           schema_pkg_apis_pipeline_v1beta1_TaskRun(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskRunInputs":                     schema_pkg_apis_pipeline_v1beta1_TaskRunInputs(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskRunList":                       schema_pkg_apis_pipeline_v1beta1_TaskRunList(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskRunOutputs":                    schema_pkg_apis_pipeline_v1beta1_TaskRunOutputs(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskRunResources":                  schema_pkg_apis_pipeline_v1beta1_TaskRunResources(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskRunResult":                     schema_pkg_apis_pipeline_v1beta1_TaskRunResult(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskRunSpec":                       schema_pkg_apis_pipeline_v1beta1_TaskRunSpec(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskRunStatus":                     schema_pkg_apis_pipeline_v1beta1_TaskRunStatus(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskRunStatusFields":               schema_pkg_apis_pipeline_v1beta1_TaskRunStatusFields(ref),
+		"./pkg/apis/pipeline/v1beta1.TaskSpec":                          schema_pkg_apis_pipeline_v1beta1_TaskSpec(ref),
+		"./pkg/apis/pipeline/v1beta1.WorkspaceBinding":                  schema_pkg_apis_pipeline_v1beta1_WorkspaceBinding(ref),
+		"./pkg/apis/pipeline/v1beta1.WorkspaceDeclaration":              schema_pkg_apis_pipeline_v1beta1_WorkspaceDeclaration(ref),
+		"./pkg/apis/pipeline/v1beta1.WorkspacePipelineTaskBinding":      schema_pkg_apis_pipeline_v1beta1_WorkspacePipelineTaskBinding(ref),
+		"./pkg/apis/resource/v1alpha1.PipelineResource":                 schema_pkg_apis_resource_v1alpha1_PipelineResource(ref),
+		"./pkg/apis/resource/v1alpha1.PipelineResourceList":             schema_pkg_apis_resource_v1alpha1_PipelineResourceList(ref),
+		"./pkg/apis/resource/v1alpha1.PipelineResourceSpec":             schema_pkg_apis_resource_v1alpha1_PipelineResourceSpec(ref),
+		"./pkg/apis/resource/v1alpha1.PipelineResourceStatus":           schema_pkg_apis_resource_v1alpha1_PipelineResourceStatus(ref),
+		"./pkg/apis/resource/v1alpha1.ResourceDeclaration":              schema_pkg_apis_resource_v1alpha1_ResourceDeclaration(ref),
+		"./pkg/apis/resource/v1alpha1.ResourceParam":                    schema_pkg_apis_resource_v1alpha1_ResourceParam(ref),
+		"./pkg/apis/resource/v1alpha1.SecretParam":                      schema_pkg_apis_resource_v1alpha1_SecretParam(ref),
 	}
 }
 
@@ -328,14 +328,14 @@ func schema_pkg_apis_pipeline_v1beta1_Cl
 					},
 					"status": {
 						SchemaProps: spec.SchemaProps{
-							Ref: ref("github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.CloudEventDeliveryState"),
+							Ref: ref("./pkg/apis/pipeline/v1beta1.CloudEventDeliveryState"),
 						},
 					},
 				},
 			},
 		},
 		Dependencies: []string{
-			"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1.CloudEventDeliveryState"},
+			"./pkg/apis/pipeline/v1beta1.CloudEventDeliveryState"},
 	}
 }
 # […]
```

Interesting, it proves to be useful aready :angel:. Seems like the generated files are different from when running on the CI vs on your machine @jinchihe ?

jinchihe commented:
I just back at work. Will take a look the problem asap. Thanks.

jinchihe commented:
@vdemeester CI test problem passed now, that's ready to go. Thanks.

jinchihe commented:
@vdemeester kindly ping, any more comments? Thanks.

vdemeester commented on hack/update-openapigen.sh:
```diff
@@ -0,0 +1,27 @@
+    --go-header-file hack/boilerplate/boilerplate.go.txt >> /dev/null
+
+echo "Generating swagger file ..."
+go run cmd/spec-gen/main.go 0.1 > pkg/apis/pipeline/v1beta1/swagger.json
```
quick question: what `0.1` means here ? :thinking:

jinchihe replied:
That's using for swagger file version, in fact, that's not very import here :-)

vdemeester replied:
ok, thanks :+1:

vdemeester reviewed (commented):
/lgtm

jinchihe commented:
Shall wee merge the PR now? if can be merged, I will update Python SDK PR accordingly. Thanks a lot! @vdemeester

afrittoli commented on pkg/apis/pipeline/v1beta1/swagger.json:
```diff
@@ -0,0 +1,2186 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "description": "Python SDK for Tekton Pipeline",
```
Is this right? If this is python specific it should probably be in the experimental repo?

jinchihe replied:
Good catch! updated to "Tekton Pipeline",  we used the swagger file to generate Python SDK, yeah, there should be a common name.

afrittoli commented on pkg/apis/pipeline/v1beta1/swagger.json:
```diff
@@ -0,0 +1,2186 @@
+  "info": {
+    "description": "Python SDK for Tekton Pipeline",
+    "title": "Tekton",
+    "version": "v0.1"
```
This version is the same specified in https://github.com/tektoncd/pipeline/pull/3081/files#diff-03086a4a73c6f26ac1efba516c7ed86fR27?

It looks like it would make sense for it to be to set to a unique value, perhaps related to the git sha it was generated from? We could extend our release process to generate these with the specific version and publish it somewhere on the `tekton.dev` domain, so that it would be browsable per-version

jinchihe replied:
Updated to the latest release version. I think we cannot use the git sha, since if so, the `diff pkg/*` in CI will be failed, so I used the latest release tag `v0.16.1`. Personally that's not critical for now if just for Python SDK generation :-)

afrittoli reviewed (commented):
Thank you for this, it's great to have this generated!
I'm also happy that existing CI job can already check that this stays in sync :) 

It would be good do have some docs about generating this. 
Also I would like to better understand the version bit and how are we going to publish this.

If I import [this](https://raw.githubusercontent.com/tektoncd/pipeline/a631f1e5dedaf7b588d18639537c2c62f4d81056/pkg/apis/pipeline/v1beta1/swagger.json) into [swagger editor](https://editor.swagger.io/) I get a bunch of errors like:
```
Semantic error at definitions.pod.Template.properties.affinity.$ref
$refs must reference a valid location in the document
Jump to line 14
```

After the errors the heading says:
![image](https://user-images.githubusercontent.com/2205608/92717935-f9b0b300-f358-11ea-8e82-4dad1d755cb4.png)

So I think it's important for description and version to be meaningful.

jinchihe commented:
@afrittoli Thanks for your reviewing. I updated the `description` and `version` for swagger file.

> It would be good do have some docs about generating this.

I already updated the hack/README.md file , and that's easy, just run the script.
https://github.com/tektoncd/pipeline/blob/dd75884a6e7ae4fc1e6d11ac515640b725665ff8/hack/README.md

> If I import this into swagger editor I get a bunch of errors like:

I tried with new one, no error happen now.

jinchihe commented:
/test pull-tekton-pipeline-build-tests

jinchihe commented:
That's strange, why the `pull-tekton-pipeline-build-tests` failed, seems there is problem while checking license?

jinchihe commented:
@afrittoli  @vdemeester The CI test is passed now. Thanks.

vdemeester reviewed (commented):
/lgtm

sbwsg commented:
/hold

sbwsg commented:
Hey sorry @jinchihe could you update the commit message to:

1. Remove mention of Python SDK
2. Describe the actual changes you've made with a couple paragraphs?

After that I think this is ready to go!

sbwsg commented:
/approve cancel

jinchihe commented:
@sbwsg so sorry... I missed your comments, update later.  I already updated that and removed SDK related message. 

And I noticed there are some changes in types, so I also re-generated the OpenAPI specification and Swagger file. Thanks a lot!

sbwsg commented:
/hold cancel

vincent-pli commented:
/lgtm
